code/modules/power/generators/turbine.dm

```diff
@@ -315,55 +308,45 @@
 		return
 	return ..()
 
-/obj/machinery/power/turbine/interact(mob/user)
-
-	if( !Adjacent(user)  || (stat & (NOPOWER|BROKEN)) && (!issilicon(user)) )
-		user.unset_machine(src)
-		user << browse(null, "window=turbine")
-		return
-
-	var/t = "<TT><B>Gas Turbine Generator</B><HR><PRE>"
-
-	t += "Generated power : [round(lastgen)] W<BR><BR>"
-
-	t += "Turbine: [round(compressor.rpm)] RPM<BR>"
-
-	t += "Starter: [ compressor.starter ? "<A href='?src=[UID()];str=1'>Off</A> <B>On</B>" : "<B>Off</B> <A href='?src=[UID()];str=1'>On</A>"]"
-
-	t += "</PRE><HR><A href='?src=[UID()];close=1'>Close</A>"
-
-	t += "</TT>"
-	var/datum/browser/popup = new(user, "turbine", name)
-	popup.set_content(t)
-	popup.open()
-
-	return
-
-/obj/machinery/power/turbine/Topic(href, href_list)
+/obj/machinery/power/turbine/attack_hand(mob/user)
+	. = ..()
+	ui_interact(user)
+
+/obj/machinery/power/turbine/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "TurbineComputer", name, 400, 300, master_ui, state)
+		ui.open()
+
+/obj/machinery/power/turbine/ui_data(mob/user)
+	var/list/data = list()
+	data["compressor"] = compressor ? TRUE : FALSE
+	data["compressor_broken"] = (!compressor || (compressor.stat & BROKEN)) ? TRUE : FALSE
+	data["turbine"] = compressor?.turbine ? TRUE : FALSE
+	data["turbine_broken"] = (!compressor || !compressor.turbine || (compressor.turbine.stat & BROKEN)) ? TRUE : FALSE
+	data["online"] = compressor?.starter
+	data["power"] = compressor?.turbine?.lastgen
+	data["rpm"] = compressor?.rpm
+	data["temperature"] = compressor?.gas_contained.return_temperature()
```

Comment:
for all of these where you're doing `compressor?.`, you can just avoid sanity checking stuff by doing an if check to see if `compressor` exists and nesting all these data[] assignments under it. If compressor is null then all of these data hashes are going to have null values regardless.

As long as you establish the `const {} = data;` stuff in your TGUI like I was talking about you shouldn't have any errors

Reply:
I believe this commit will resolve this one. [e8fbc69](https://github.com/ParadiseSS13/Paradise/pull/21226/commits/e8fbc696974dca6673380b5bb07645ca6ed05a8c)